Sync static pods from Kubelet to the API server

yujuhong opened the pull request:
Currently, API server is not aware of the static pods (manifests from
sources other than the API server, e.g. file and http) at all. This is
inconvenient since users cannot check the static pods through kubectl.
It is also suboptimal because scheduler is unaware of the resource
consumption by these static pods on the node.

This change syncs the information back to the API server by creating a
mirror pod via API server for each static pod.
- Kubelet creates containers for the mirror pod, instead of the static
  pod, which is filtered out before SyncPods.
- If a mirror pod gets deleted, Kubelet will re-create one. The
  containers will also be re-created after syncing the new mirror pod.
- If a static pod gets removed from the source (e.g. manifest file
  removed from the directory), the orphaned mirror pod will be deleted
  by Kubelet.

This fixes #4090.

yujuhong committed:
Sync static pods from Kubelet to the API server

Currently, API server is not aware of the static pods (manifests from
sources other than the API server, e.g. file and http) at all. This is
inconvenient since users cannot check the static pods through kubectl.
It is also sub-optimal because scheduler is unaware of the resource
consumption by these static pods on the node.

This change syncs the information back to the API server by creating a
mirror pod via API server for each static pod.

 - Kubelet creates containers for the static pod, as it would do
   normally.

 - If a mirror pod gets deleted, Kubelet will re-create one. The
   containers are sync'd to the static pods, so they will not be
   affected.

 - If a static pod gets removed from the source (e.g. manifest file
   removed from the directory), the orphaned mirror pod will be deleted.

Note that because events are associated with UID, and the mirror pod has
a different UID than the original static pod, the events will not be
shown for the mirror pod when running `kubectl describe pod
<mirror_pod>`.

yujuhong commented:
@dchen1107 and @bgrant0607, This PR is almost complete, but I would like some feedback, especially on whether kubelet should sync to the static pod or the mirror pod.

Originally, I went with the former (syncing the static pod), but I realized `kubectl describe pod <mirror_pod>` returned the status of the pod, but not the events of the pod. This is because we records events using the UID, and static and mirror pods have, of course, different UIDs.

Then, I switched to let kubelet sync with the mirror pods (implemented in this PR). The cons are:
  - If you delete the mirror pod, kubelet will kill the current containers for the pod, recreate a mirror pod, and start a new set of containers. This may be undesirable, but it still reasonable because pods are not expected to be durable.
  - The other downside, which may be more important is, I am not sure if this model will work well with the plan to start individual master component as pods. Kubelet will need to talk to the Apiserver to create mirror pods for the component first.

Thoughts?

vmarmol commented on pkg/kubelet/kubelet.go:
```diff
@@ -1573,26 +1573,130 @@ func (kl *Kubelet) syncLoop(updates <-chan PodUpdate, handler SyncHandler) {
 
-// Updated the Kubelet's internal pods with those provided by the update.
+func isStaticPod(pod *api.BoundPod) bool {
+	if getPodSource(pod) != ApiserverSource {
```
What options here?

vmarmol replied:
Sorry, I should describe that more :) What are the options for pod source? We have HTTP, static, and API server. Are we ignoring HTTP here as well?

vmarmol replied:
Nvm this, if it's not API server it is static...

vmarmol commented on pkg/kubelet/kubelet.go:
```diff
@@ -1573,26 +1573,130 @@ func (kl *Kubelet) syncLoop(updates <-chan PodUpdate, handler SyncHandler) {
+	}
+}
+
+func (kl *Kubelet) createMirrorPod(pod *api.BoundPod) error {
```
We should probably batch all the creation of mirror pods and do them async on the podWorkers. Doing these here will slow down pod startup latency overall. Maybe we can mark the pods as mirror pods and let syncPod() handle the create? Or just remove all static pods from here and only act on them when they're served by the API server

It can be argued that they don't happen often and shouldn't affect operation though.

vmarmol commented on pkg/kubelet/kubelet.go:
```diff
@@ -1573,26 +1573,130 @@ func (kl *Kubelet) syncLoop(updates <-chan PodUpdate, handler SyncHandler) {
+	numPodsByName := make(map[string]int)
+	for _, pod := range pods {
+		name := GetPodFullName(&pod)
+		if num, ok := numPodsByName[name]; ok {
```
nit: `numPodsByName[name]++` will do the right thing. It it exists it increments it and if it doesn't it will create it (with default value 0) and increment it.

vmarmol commented:
I like how well this works and how conceptually simple it is :) A downside is that the Kubelet won't be able to start any static pods while the API server is down. Depending on what we run as a static pod, this may be an issue

yujuhong commented:
@vmarmol, that's why I mentioned in my previous comment that there were two ways to do it :)  If we go with the first option, which is syncing the static pods, then the lack of API server won't affect anything. Unfortunately, the events will be missing from the output because of the mismatching UID. What do you think?

dchen1107 commented:
Haven't read the PR in detail yet, but related to dependency on apiserver to start any static pods, I want to say NO. This will definitely break:

1) We want to use kubelet to run master components as POD. 
2) We still have internal users running standalone containervm to start pods, which doesn't have apiserver at all.

bgrant0607 commented:
I agree with @dchen1107.

I think we should create a way for Kubelet to write the metadata fields, such as uid and creationTimestamp, for its pods.

Maybe @smarterclayton could advise how we could bypass FillObjectMetaSystemFields:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/rest/create.go#L61

smarterclayton commented:
Good question - eventually someone comes up with a use case to let an automated system (importer/migrator) set both uid and creationTimestamp.  It's a system level action that is not in the normal purview of users.  We'd have to distinguish between allowed and not allowed - our way to handle that today is a different resource endpoint. 

> On Mar 12, 2015, at 6:30 PM, Brian Grant <notifications@github.com> wrote:
> 
> I agree with @dchen1107.
> 
> I think we should create a way for Kubelet to write the metadata fields, such as uid and creationTimestamp, for its pods.
> 
> Maybe @smarterclayton could advise how we could bypass FillObjectMetaSystemFields:
> https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/rest/create.go#L61
> 
> —
> Reply to this email directly or view it on GitHub.
>

yujuhong commented:
@vmarmol, FYI, I uploaded a new patch syncing to the static pods. I will be working on pushing mirror pod creation/deletion to the pod workers next.

vmarmol commented:
New logic looks good, I have a comment for general flow which if you agree with please make :)

WDYT of having the handling of mirror nodes be done by a small object we inject into the Kubelet? (mirrorManager? I am terrible at names...) Similar to what we do today for podWorkers, imageManager, cAdvisor, etc. It seems logically separate enough to be able to be its own structure. This will make it easier to test, reason about, and reduce the size of kubelet.go. I've been trying to slowly push new changes that route.

yujuhong commented:
@vmarmol, sure, that sounds reasonable.

As for your suggestion about creating/deleting mirror pods in per-pod threads (go routines), I digged a little bit more and here is what I found. 

 - We can certainly create the pod in syncLoop using per-pod threads.
 - However, the only way we can ensure that the mirror pods are garbage collected is to delete them in SyncPods(), similarly to how we GC the containers. The main reason is that deletion may fail, and you can't really hold the thread hostage until the deletion succeeds. The minor reason (which could be fixed) is that a per-pod thread expects a pod update which contains a pod. The deletion, nonetheless, lacks the pod and require to cache (?) the previous pod somewhere for the update. Even if we go with this, we'd have to tolerate the possible race condition between mirror deletion and creation (which might be okay).

What do you think about this, compared with the current implementation in the PR?

vmarmol commented:
The plan sounds good and should affect pod startup less than the current approach. We may have to re-architect podWorker a bit for this, but I think it's doable. The create and delete cases run into the same issues because both operations can fail and take their time. If we do them at the end of each syncPod() we should save ourselves from most of the trouble. We'll need to timeout and retry in case the apiserver is unavailable.

I can see how the delete case would be easier to handle in SyncPods(). I'm okay starting with that and moving it into the pod worker in the future. If we do the delete after all the syncPod()s (as we do today for container deletion), then we should be okay since it won't affect startup time. I think long-term we'll have the pod workers doing the deletion of containers in the pod and we'll be able to do the deletion of the mirrors there too.

The create/delete race probably won't be bad since the create will come with a new UID and in our minds it's a different pod. We'd only create when the pod was new to us and only delete when it was forgotten.

satnam6502 commented:
If by static pods you also mean the pods for logging that get laid down on each node via a manifest file and Salt -- then I think I like this because it gives me a way to answer the question "what cluster level logging do we have?" by looking to see what logging pods are actually running.

yujuhong commented:
@vmarmol, create is allowed to fail, which is fine. As long as the pod is still alive, syncPod() will run and try again. I don't think we need to retry in this case, do we? I agree that if we rework pod workers, we may be able to do the deletion, we can move towards that direction later :)

As for the race condition, I'd like to point out that we don't care about the UID for mirror pods (at least not for now), so the fact that it's a new UID doesn't really mean anything. We delete the mirror pods by pod name+namespace.

@satnam6502, if you mean the fluentd pods, then yes :)
With this PR, you should be able to see their status using kubectl. The events are not associated with the pod yet though, which will be fixed in the (near?) future.

vmarmol commented:
@yujuhong good point, that form of create retry SGTM. 

Oooh, you've been trying to tell me about that race, sorry I haven't understood it till now :)  

Just merged some refactoring to syncPod() btw.

yujuhong commented:
@vmarmol, I uploaded a new patch using per-pod workers to create mirror pods. I also moved some mirror pod-related functions to a new file (mirror_manager.go). 

This PR will need to be rebased after BoundPod has been removed (#5433).

vmarmol commented:
#5433 just went in btw, taking a look at this PR now.

vmarmol commented on pkg/kubelet/kubelet.go:
```diff
@@ -78,19 +78,23 @@ const (
 	// type of sync is occuring per pod. StartTime specifies the time at which
 	// syncing began (for use in monitoring).
-	SyncPods(pods []api.BoundPod, podSyncTypes map[types.UID]metrics.SyncPodType, startTime time.Time) error
+	SyncPods(pods []api.BoundPod, podSyncTypes map[types.UID]metrics.SyncPodType, mirrorPods util.StringSet,
```
Can we document what mirrorPods contains and whether they are filtered from pods?

yujuhong replied:
The details of mirror pods are documented in mirror_manager. I will add a comment in `Kubelet.mirrorPods` to point to that file, as well as the fact that they are filtered out from `Kubelet.pods`.

vmarmol commented on pkg/kubelet/kubelet.go:
```diff
@@ -1416,6 +1428,11 @@ func (kl *Kubelet) syncPod(pod *api.BoundPod, containersInPod dockertools.Docker
 
+	// Create a mirror pod for the static pod if it doesn't exist yet.
+	if !hasMirrorPod && isStaticPod(pod) {
+		kl.mirrorManager.CreateMirrorPod(pod, kl.hostname)
```
Lets log the failure here and remove the logging from CreateMirrorPod() that way we only need to do it once.

vmarmol commented on pkg/kubelet/kubelet.go:
```diff
@@ -1723,6 +1743,7 @@ func (kl *Kubelet) updatePods(u PodUpdate, podSyncTypes map[types.UID]metrics.Sy
 	switch u.Op {
 	case SET:
 		glog.V(3).Infof("SET: Containers changed")
+		newPods, newMirrorPods := filterAndCategorizePods(u.Pods)
```
WDYT of making this a member of mirrorManager?

yujuhong replied:
No, I prefer not to. I explicitly made it not a member because it's guarded by the same lock as kubelet.pods, and they have to be in sync.

A more reasonable approach would be to have a pod manager that encompasses both (regular and mirror pods). That'd be yet another large kubelet refactoring, which I'd like to avoid in the current PR.

vmarmol replied:
SGTM, the longer-term podManager sounds like the right approach.

vmarmol commented on pkg/kubelet/mirror_manager.go:
```diff
@@ -0,0 +1,165 @@
+	"github.com/golang/glog"
+)
+
+// Kubelet discover pod updates from 4 sources: file, http, etcd, and
```
nit: etcd is no longer a source.

yujuhong replied:
Thanks! Will remove that. Our codebase is changing so fast that it's hard to keep track :|

vmarmol commented on pkg/kubelet/mirror_manager.go:
```diff
@@ -0,0 +1,165 @@
+
+// Creates a mirror pod.
+func (self *basicMirrorManager) CreateMirrorPod(pod *api.BoundPod, hostname string) error {
+	if self.apiserverClient == nil {
```
I'd say to assume this is always true and not check it within the member methods. It'll get tricky to track what we support when what dependencies are/are not available.

yujuhong replied:
We need to support standalone kubelets, don't we? Mirror pods are an add-on, which should not affect kubelet's basic functions. Or were you suggesting that I moved the check elsewhere?

vmarmol replied:
Ah, that's a good point. This is fine as is.

vmarmol commented on pkg/kubelet/mirror_manager.go:
```diff
@@ -0,0 +1,165 @@
+}
+
+// Delete all orphaned mirror pods.
+func (kl *Kubelet) deleteOrphanedMirrorPods(pods []api.BoundPod, mirrorPods util.StringSet) {
```
We can probably make this a member of mirrorManager and have Kubelet call it since it doesn't use any special Kubelet state.

A pet peeve of mine is also to keep all of a struct's methods in one file, but that's just me :)

yujuhong replied:
How about I make it a function in mirror_manager.go, but not a method of either Kubelet or mirrorManager? This way we can easily test the function without having to implement it in fakeMirrorManager at all.

Sorry for having a Kubelet method here. I saw many precedence in files under pkg/kubelet, and thought this was an established pattern ;)

vmarmol replied:
I feel like it still belong in mirrorManager since it is a mirror-based util. We'd still have to fake mirrorManager today since the deleteOrphaned() will call Delete(). I'm not blocking on this if you feel strongly however.

No worries, I'm trying to start the trend against that precedence!

vmarmol replied:
After reading through your unit tests I can see why you wanted to have it be a Kubelet method. I'm fine going with that if you prefer. 

Sorry for the churn!

vmarmol commented on pkg/kubelet/mirror_manager.go:
```diff
@@ -0,0 +1,165 @@
+			return source, nil
+		}
+	}
+	return "", errors.New(fmt.Sprintf("Cannot get source of pod %", pod.UID))
```
nit: use `fmt.Errorf()` it does exactly what you want here :) also, errors start with lowercase by Go convention. Here and elsewhere

yujuhong replied:
Thanks for the tip. Will fix in the next patch/

vmarmol commented on pkg/kubelet/mirror_manager.go:
```diff
@@ -0,0 +1,165 @@
+
+func isStaticPod(pod *api.BoundPod) bool {
+	source, err := getPodSource(pod)
+	if err == nil && source != ApiserverSource {
```
nit: `return err == nil && source != ApiserverSource` unless you think its less readable

yujuhong replied:
Done.

vmarmol commented on pkg/kubelet/mirror_manager.go:
```diff
@@ -0,0 +1,165 @@
+	}
+}
+
+func parsePodFullName(podFullName string) (string, string, error) {
```
Can we move this next to GetPodFullName() so it can be re-used by others?

Also +1 to having an error as we learned with the Docker one :D

yujuhong replied:
I actually hid the function here to discourage people from using it, since most of the kubelet code is driving away from using full names. I'll move it to types.go anyway in case someone else try re-creating it again :)

vmarmol commented on pkg/kubelet/mirror_manager.go:
```diff
@@ -0,0 +1,165 @@
+func parsePodFullName(podFullName string) (string, string, error) {
+	parts := strings.Split(podFullName, "_")
+	if len(parts) != 2 {
+		glog.Error("Failed to parse a pod full name %q", podFullName)
```
Errorf (missing the "f")

vmarmol commented on pkg/kubelet/pod_workers.go:
```diff
@@ -60,11 +60,15 @@ type workUpdate struct {
 	pod *api.BoundPod
 
+	// Whether there exists a mirror pod for pod.
+	hasMirrorPod bool
```
WDYT: isMirrorPod?

yujuhong replied:
I think that'd be slightly confusing. None of the pods we are syncing are actually mirror pods......but they may have mirror pods.

vmarmol replied:
Aaaah I see what you mean, yes that makes sense.

yujuhong commented:
Thanks for the reminder!

Note to self: rebase again if #5205 lands before this PR (which is most likely the case).

yujuhong commented:
@vmarmol, I've rebased (to include the BoundPod-> Pod PR), and uploaded a new patch that addressed your comments. Added a new commit to make the change more obvious. Will squash after the review is done.

vmarmol commented:
#5205 just went in (sorry again :P)

vmarmol commented:
Looks good, only minor things.

vmarmol commented on pkg/kubelet/mirror_manager_test.go:
```diff
@@ -0,0 +1,214 @@
+	}
+}
+
+func TestCreateMirrorPod(t *testing.T) {
```
These last two tests I think should be in kubelet_test.go since they test Kubelet functionality. We can probably make the last one call garbageCollect directly if we'd like, but I do see the value of testing the Kubelet path which is why you wrote it this way.

yujuhong replied:
Done!

yujuhong commented:
Uploaded a new patch which addresses previous comments. Also squashed the additional commits.

I was half way through rebasing #5205, and finally noticed that it was reverted :\

vmarmol commented on pkg/kubelet/kubelet.go:
```diff
@@ -235,6 +241,11 @@ type Kubelet struct {
 	pods    []api.Pod
+	// Record the set of mirror pods (see mirror_manager.go for more details);
+	// similar to pods, this is not immutable and is protected by the same podLock.
+	// Note that Kubelet.pods do not containe mirror pods as they are filtered
```
typo: "containe"

vmarmol commented:
Yeah that was unfortunate :( but lets get this in before the reincarnation of that PR comes in. LGTM, just one typo. Have you had a chance to run e2e?

yujuhong commented:
Typo corrected.

I ran e2e once but there was some seemingly unrelated failures. Don't wanna take any chances, so I am going to run e2e again. Will update once it passes (or if I can reliably rule out unrelated flakes).

vmarmol commented:
@yujuhong mentioned the e2e's passing. LGTM, merging. Thanks!
